packages/@aws-cdk/aws-servicecatalogappregistry/lib/application.ts

```diff
@@ -42,6 +56,54 @@ export interface ApplicationProps {
 abstract class ApplicationBase extends cdk.Resource implements IApplication {
   public abstract readonly applicationArn: string;
   public abstract readonly applicationId: string;
+  private readonly associatedAttributeGroups: Set<string> = new Set();
+  private readonly associatedResources: Set<string> = new Set();
+
+  /**
+   * Associate an attribute group with application
+   * If the attribute group is already associated, it will ignore duplicate request.
+   */
+  public associateAttributeGroup(attributeGroup: IAttributeGroup): void {
+    if (!this.associatedAttributeGroups.has(attributeGroup.node.addr)) {
+      const hashId = this.generateUniqueHash(attributeGroup.node.addr);
+      new CfnAttributeGroupAssociation(this, `AttributeGroupAssociation${hashId}`, {
+        application: this.applicationId,
+        attributeGroup: attributeGroup.attributeGroupId,
+      });
+      this.associatedAttributeGroups.add(attributeGroup.node.addr);
+    }
+  }
+
+  /**
+   * Associate a stack with the application
+   * If the resource is already associated, it will ignore duplicate request.
+   * A stack can only be associated with one application.
+   */
+  public associateStack(stack: cdk.Stack): void {
+    if (!this.associatedResources.has(stack.node.addr)) {
+      const hashId = this.generateUniqueHash(stack.node.addr);
+      new CfnResourceAssociation(this, `ResourceAssociation${hashId}`, {
+        application: this.applicationId,
+        resource: stack.stackId,
+        resourceType: 'CFN_STACK',
+      });
+      this.associatedResources.add(stack.node.addr);
+    }
+  }
+
+  /**
+   * Create a unique id
+   */
+  protected abstract generateUniqueHash(resourceAddress: string): string;
+}
+
+/**
+ * Generates a unique hash identfifer using SHA256 encryption algorithm
+ */
+export function hashValues(...values: string[]): string {
```

Comment:
No need for `export` here, right?

Also, can you move this function to the end of this file?

Reply:
yes, you're right.  Also I had put it here because I thought it should go above the first time it's called.  But I think that guidance might just be for classes that are utilized, and things like generic functions and enums should be at the end?

Reply:
The guidance is: public elements of the file (customer-facing) go first, ideally before they are used.

All private elements (not customer-facing) are at the end of the life, after they've been used.

Reply:
got it, thanks for clarifying